File: src/main/java/me/reckter/telegram/listener/ReflectionListener.java
package me.reckter.telegram.listener;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import me.reckter.telegram.model.GroupChat;
import me.reckter.telegram.model.Message;
import me.reckter.telegram.model.User;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;


/**
 * @author hannes
 */
public class ReflectionListener implements GroupCommandListener, GroupMessageListener, UserMessageListener, UserCommandListener {

    private static Logger Log = LoggerFactory.getLogger(ReflectionListener.class)

    List<Method> commandListener = new ArrayList<>();
    List<Method> messageListener = new ArrayList<>();
    List<Method> locationListener = new ArrayList<>();
    List<Method> userCreateListener = new ArrayList<>();
    List<Method> userLeaveListener = new ArrayList<>();

    Object listenerObject;


    public ReflectionListener(Object listener) {
        this.listenerObject = listener;
        parse();
    }

    private void parse() {
        Method[] methods = listenerObject.getClass().getMethods();
        for(Method method : methods) {
            if(method.isAnnotationPresent(OnCommand.class)) {
                commandListener.add(method);
            }
            if(method.isAnnotationPresent(OnMessage.class)) {
                messageListener.add(method);
            }
            if(method.isAnnotationPresent(OnLocation.class)) {
                locationListener.add(method);
            }
            if(method.isAnnotationPresent(OnUserJoin.class)) {
                userCreateListener.add(method);
            }
            if(method.isAnnotationPresent(OnUserLeave.class)) {
                userLeaveListener.add(method);
            }
        }
    }


    private List<String> getArguments(String text) {

        text = text.substring(1);
        text = text + " ";
        String[] split = text.split(" ", 2);
        text = split[0].replace("_", " ") + " " + split[1];

        List<String> ret = new ArrayList<>();

        int argumentStartedAt = 0;
        boolean insideQoutes = false;
        for(int i = 0; i < text.length(); i++) {

            if(text.charAt(i) == '\\') {
                i++;
                continue;
            }

            if(text.charAt(i) == '"') {
                insideQoutes = !insideQoutes;
            }

            if(!insideQoutes && text.charAt(i) == ' ') {
                ret.add(text.substring(argumentStartedAt, i));
                argumentStartedAt = i + 1;
            }
        }
        return ret;
    }


    private void invokeSimpleHandler(Method method, Message message) {
        method.setAccessible(true);
        try {
            method.invoke(listenerObject, message);
        } catch(IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException("There seems to be a Problem with the Handler!", e);
        }
    }

    private boolean invoke(Message message) {
        List<Method> methods;

        if(message.getLocation() != null) {
            locationListener.stream().forEach(method -> invokeSimpleHandler(method, message));
        } else if(message.getNewChatParticipant() != null) {
            userCreateListener.stream().forEach(method -> invokeSimpleHandler(method, message));
        } else if(message.getLeftChatParticipant() != null) {
            userLeaveListener.stream().forEach(method -> invokeSimpleHandler(method, message));
        } else if(message.text != null && message.text.startsWith("/")) {

            List<String> arguments = getArguments(message.text);
            String commandCalled = arguments.get(0);


            final boolean[] foundCommand = {false};
            commandListener.stream().filter(method ->
                                                    Stream.of(method.getDeclaredAnnotation(OnCommand.class).value())
                                                            .anyMatch(name -> name.toLowerCase().equals(commandCalled.toLowerCase())))
                    .forEach(method -> {
                        OnCommand onCommand = method.getDeclaredAnnotation(OnCommand.class);
                        if(onCommand.type() == ChatType.ALL
                                || onCommand.type() == ChatType.GROUP && message.chat instanceof GroupChat
                                || onCommand.type() == ChatType.USER && message.chat instanceof User) {

                            foundCommand[0] = true;
                            try {
                                method.setAccessible(true);
                                method.invoke(listenerObject, message, arguments);
                            } catch(IllegalAccessException | InvocationTargetException e) {
                                throw new RuntimeException("There seems to be a Problem with the Handler!", e);
                            }
                        }
                    });
            if(!foundCommand[0]) {
                return false;
            }


        } else if(message.text != null) {
            messageListener.stream().forEach(method -> {
                OnMessage onMessage = method.getDeclaredAnnotation(OnMessage.class);

                Matcher matcher = Pattern.compile(onMessage.regex()).matcher(message.text);

                if(!matcher.matches()) {
                    return;
                }

                Map<String, String> groups = new HashMap<>();
                for(String group : onMessage.matchingGroup()) {
                    groups.put(group, matcher.group(group));
                }
                try {
                    method.invoke(listenerObject, message, groups);
                } catch(IllegalAccessException | InvocationTargetException e) {
                    e.printStackTrace();
                }
            });
        } else {
            ObjectMapper objectMapper = new ObjectMapper();
            try {
                Log.info("could not find suitable action for " + objectMapper.writeValueAsString(message));
            } catch(JsonProcessingException e) {
                e.printStackTrace();
            }
        }

        return true;
    }

    @Override
    public boolean onUserCommand(Message message) {
        return invoke(message);
    }

    @Override
    public void onGroupMessage(Message message) {
        invoke(message);
    }

    @Override
    public void onUserMessege(Message message) {
        invoke(message);
    }

    @Override
    public boolean onGroupCommand(Message message) {
        return invoke(message);
    }
}

File: src/main/java/me/reckter/telegram/listener/ReflectionListener.kt
package me.reckter.telegram.listener

import me.reckter.telegram.model.Message
import me.reckter.telegram.model.MessageType
import me.reckter.telegram.model.update.CallbackQuery
import java.lang.reflect.Method

/**
 *  @author Hannes Güdelhöfer
 */
class ReflectionListener(val listener: Any) : MessageListener, CommandListener, EditListener, LocationListener, CallBackListener {


    val methods = mutableMapOf<MessageType, MutableList<Method>>()

    val callBackListener = mutableListOf<Method>()

    init {
        MessageType.values()
                .filter { it != MessageType.ALL }
                .forEach {
                    methods.put(it, mutableListOf<Method>())
                }
        parse()
    }

    private fun parse() {
        val listenerMethods = listener.javaClass.methods
        for (listenerMethod in listenerMethods) {
            val messageTypes: List<MessageType> = when {
                listenerMethod.isAnnotationPresent(OnMessage::class.java) -> {
                    var tmp = listenerMethod.getAnnotation(OnMessage::class.java).value.toMutableSet()
                    if (tmp.any { it == MessageType.ALL }) {
                        tmp.addAll(MessageType.values())
                    }
                    tmp.remove(MessageType.ALL)
                    tmp.toList()
                }
                listenerMethod.isAnnotationPresent(OnCommand::class.java) -> listOf(MessageType.COMMAND)
                listenerMethod.isAnnotationPresent(OnLocation::class.java) -> listOf(MessageType.LOCATION)
                listenerMethod.isAnnotationPresent(OnUserJoin::class.java) -> listOf(MessageType.NEW_CHAT_PARTICIPANT)
                listenerMethod.isAnnotationPresent(OnUserLeave::class.java) -> listOf(MessageType.LEFT_CHAT_PARTICIPANT)
                listenerMethod.isAnnotationPresent(OnCallBack::class.java) -> {
                    listenerMethod.isAccessible = true
                    callBackListener.add(listenerMethod)
                    listOf()
                }
                else -> {
                    listOf()
                }
            }
            if (messageTypes.size > 0) {
                listenerMethod.isAccessible = true
                messageTypes.forEach {
                    methods[it]!!.add(listenerMethod)
                }
            }
        }
    }


    private fun methodAcceptsCommand(method: Method, command: String): Boolean {
        if(!method.isAnnotationPresent(OnCommand::class.java)) {
            return true
        }
        val trigger = method.getAnnotation(OnCommand::class.java)

        return trigger.value.any { it == command }
    }


    override fun onEdit(message: Message) {
        methods[message.type]!!.forEach { it.invoke(listener, message) }
    }

    override fun onCommand(message: Message, arguments: List<String>) {
        methods[message.type]!!.filter { methodAcceptsCommand(it, arguments.first()) }
                .forEach {
                    if(it.parameterCount == 2) {
                        it.invoke(listener, message, arguments)
                    } else {
                        it.invoke(listener, message)
                    }
                }
    }

    override fun onMessage(message: Message) {
        methods[message.type]!!.forEach { it.invoke(listener, message) }
    }

    override fun onLocation(message: Message) {
        methods[message.type]!!.forEach { it.invoke(listener, message) }
    }

    override fun accepts(command: String): Boolean {
        return methods[MessageType.COMMAND]!!.any { methodAcceptsCommand(it, command) }
    }

    override fun OnCallBack(callbackQuery: CallbackQuery) {
        callBackListener.forEach { it.invoke(listener, callbackQuery) }
    }
}